some bug in http or whatever

master branch, but some configs maybe outdated, it's moving too fast to keep track of everything.
```
[NOTICE] [-] [core] Listening on :::2525
[NOTICE] [-] [core] Listening on :::2587
[DEBUG] [-] [server] running init_master hooks
[DEBUG] [-] [server] running init_master hook in redis plugin
[NOTICE] [-] [core] Listening on 127.0.0.1:8010
[DEBUG] [-] [server] running init_http hooks
[CRIT] [-] [core] Error: We are already running hooks! Fatal error!
[CRIT] [-] [core]     at Object.plugins.run_hooks (/storage/personal/work/coding/ruby/nodejs/node_modules/Haraka/plugins.js:233:15)
[CRIT] [-] [core]     at registerRoutes (/storage/personal/work/coding/ruby/nodejs/node_modules/Haraka/server.js:283:17)
[CRIT] [-] [core]     at done (/storage/personal/work/coding/ruby/nodejs/node_modules/Haraka/node_modules/async/lib/async.js:132:19)
[CRIT] [-] [core]     at /storage/personal/work/coding/ruby/nodejs/node_modules/Haraka/node_modules/async/lib/async.js:32:16
[CRIT] [-] [core]     at Server.<anonymous> (/storage/personal/work/coding/ruby/nodejs/node_modules/Haraka/server.js:270:13)
[CRIT] [-] [core]     at Server.emit (events.js:92:17)
[CRIT] [-] [core]     at net.js:1055:10
[CRIT] [-] [core]     at process._tickCallback (node.js:419:13)
[CRIT] [-] [core]     at Function.Module.runMain (module.js:499:11)
[CRIT] [-] [core]     at startup (node.js:119:16)
[NOTICE] [-] [core] Shutting down
```

````
[root@haraka] # cd /usr/local/lib/node_modules/
[root@haraka] # git clone https://github.com/baudehlo/Haraka.git
Cloning into 'Haraka'...
<snip>
Checking connectivity... done.
[root@haraka] # cd Haraka
[root@haraka] # npm install
<snip lots of output>
[root@haraka] # haraka -i /usr/local/haraka
[root@haraka] # cd /usr/local/haraka
[root@haraka] # sed -i .bak -e 's/^;listen=\[.*$/listen=127.0.0.9:25,127.0.0.9:465,127.0.0.9:587/' config/smtp.ini
[root@haraka] # sed -i .bak -e 's/; listen=\[::\]:80/listen=127.0.0.9:80/' config/http.ini
[root@haraka] # haraka -c /usr/local/haraka
loglevel: LOGPROTOCOL
Starting up Haraka version 2.6.1
[INFO] [-] [core] Loading plugins
[INFO] [-] [core] Loading plugin: access
[DEBUG] [-] [core] no timeout in access.timeout
[DEBUG] [-] [core] no timeout in plugin_timeout
[DEBUG] [-] [core] plugin access timeout is: 30s
[INFO] [-] [core] loaded 7185 Public Suffixes
<snip>
[DEBUG] [-] [core] registered hook connect to max_unrecognized_commands.hook_connect
[DEBUG] [-] [core] registered hook unrecognized_command to max_unrecognized_commands.hook_unrecognized_command
[ERROR] [-] [core] Missing tls_key.pem for port 465
[NOTICE] [-] [core] Listening on 127.0.0.9:25
[NOTICE] [-] [core] Listening on 127.0.0.9:587
[DEBUG] [-] [server] running init_master hooks
[INFO] [-] [core] [outbound] Loading outbound queue from /usr/local/haraka/queue
[NOTICE] [-] [core] Listening on 127.0.0.9:80
[DEBUG] [-] [server] running init_http hooks
[DEBUG] [-] [server] running init_wss hooks
[INFO] [-] [core] using html docroot: /usr/local/haraka/html
[DEBUG] [-] [dnsbl] looking up: 2.0.0.127.zen.spamhaus.org.
^C
[NOTICE] [-] [core] SIGINT received
[NOTICE] [-] [core] Shutting down
[root@haraka] # echo 'watch' >> config/plugins 
[root@haraka] # haraka -c /usr/local/haraka
loglevel: LOGPROTOCOL
Starting up Haraka version 2.6.1
[INFO] [-] [core] Loading plugins
<snip>
[DEBUG] [-] [server] running init_master hooks
[INFO] [-] [core] [outbound] Loading outbound queue from /usr/local/haraka/queue
[NOTICE] [-] [core] Listening on 127.0.0.9:80
[DEBUG] [-] [server] running init_http hooks
[DEBUG] [-] [server] running init_http hook in watch plugin
[DEBUG] [-] [server] hook=init_http plugin=watch function=hook_init_http params="undefined" retval=CONT msg=""
[DEBUG] [-] [server] running init_wss hooks
[DEBUG] [-] [server] running init_wss hook in watch plugin
[DEBUG] [-] [server] hook=init_wss plugin=watch function=hook_init_wss params="undefined" retval=CONT msg=""
[INFO] [-] [core] using html docroot: /usr/local/haraka/html
[DEBUG] [-] [dnsbl] looking up: 2.0.0.127.zen.spamhaus.org.
````

Works as expected here. I'd recommend starting with a stock Haraka and then introducing your customizations, one at a time until you figure out where the issue lies.

--- a/server.js
+++ b/server.js
@@ -41,7 +41,8 @@ Server.load_smtp_ini = function () {
 };
 
 Server.load_http_ini = function () {
-    Server.http_cfg = config.get('http.ini', function () {
+    Server.http = {};
+    Server.http.cfg = config.get('http.ini', function () {
         Server.load_http_ini();
     }).main;
 };
@@ -128,7 +129,6 @@ Server.createServer = function (params) {
     if (!cluster || !c.nodes) {
         Server.daemonize(c);
         Server.setup_smtp_listeners(plugins, 'master', inactivity_timeout);
-        Server.setup_http_listeners(plugins, 'master');
         return;
     }
 
@@ -138,7 +138,6 @@ Server.createServer = function (params) {
     // Cluster Workers
     if (!cluster.isMaster) {
         Server.setup_smtp_listeners(plugins, 'child', inactivity_timeout);
-        Server.setup_http_listeners(plugins, 'child');
         return;
     }
 
@@ -233,31 +232,26 @@ Server.setup_smtp_listeners = function (plugins, type, inactivity_timeout) {
     async.each(listeners, setupListener, runInitHooks);
 };
 
-Server.setup_http_listeners = function (plugins, type) {
-    if (!Server.http_cfg) return;
-    if (!Server.http_cfg.listen) return;
+Server.setup_http_listeners = function () {
+    if (!Server.http.cfg) return;
+    if (!Server.http.cfg.listen) return;
 
-    var listeners = Server.get_listen_addrs(Server.http_cfg, 80);
+    var listeners = Server.get_listen_addrs(Server.http.cfg, 80);
     if (!listeners.length) return;
 
-    var express;
     try {
-        express = require('express');
+        Server.http.express = require('express');
+        logger.loginfo('express loaded at Server.http.express');
     }
     catch (err) {
         logger.logerror('express failed to load. No http server. ' +
                 ' Try installing express with: npm install -g express');
         return;
     }
-    var app = express();
-    var server, wss;
 
-    Server.init_http_respond = function () {
-        // logger.loginfo(arguments);
-    };
-    Server.init_wss_respond = function () {
-        // logger.loginfo(arguments);
-    };
+    var app = Server.http.express();
+    Server.http.app = app;
+    logger.loginfo('express app is at Server.http.app');
 
     var setupListener = function (host_port, cb) {
         var hp = /^\[?([^\]]+)\]?:(\d+)$/.exec(host_port);
@@ -265,17 +259,20 @@ Server.setup_http_listeners = function (plugins, type) {
             return cb(new Error('Invalid format for listen in http.ini'));
         }
 
-        server = require('http').createServer(app);
+        Server.http.server = require('http').createServer(app);
 
-        server.on('listening', function () {
+        Server.http.server.on('listening', function () {
             var addr = this.address();
             logger.lognotice('Listening on ' + addr.address + ':' + addr.port);
             cb();
         });
 
-        server.on('error', function (e) { cb(e); });
+        Server.http.server.on('error', function (e) {
+            logger.logerror(e);
+            cb(e);
+        });
 
-        server.listen(hp[2], hp[1]);
+        Server.http.server.listen(hp[2], hp[1]);
     };
 
     var registerRoutes = function (err) {
@@ -283,21 +280,8 @@ Server.setup_http_listeners = function (plugins, type) {
             logger.logerror('Failed to setup http routes: ' + err.message);
         }
 
-        plugins.run_hooks('init_http', Server, app);
-
-        var WebSocketServer;
-        try { WebSocketServer = require('ws').Server; }
-        catch (e) {
-            logger.logerror('unable to load ws.\ndid you: npm install -g ws?');
-        }
-
-        if (WebSocketServer) {
-            wss = new WebSocketServer({ server: server });
-            plugins.run_hooks('init_wss', Server, wss);
-        }
-
         app.get('/plugins', Server.http_plugins);
-        app.use(express.static(Server.get_http_docroot()));
+        app.use(Server.http.express.static(Server.get_http_docroot()));
         app.use(Server.handle404);
     };
 
@@ -317,6 +301,8 @@ Server.init_master_respond = function (retval, msg) {
     // Load the queue if we're just one process
     if (!(cluster && c.nodes)) {
         out.load_queue();
+        Server.setup_http_listeners();
+        plugins.run_hooks('init_http', Server);
         return;
     }
 
@@ -329,8 +315,7 @@ Server.init_master_respond = function (retval, msg) {
         }
         Server.daemonize();
         // Fork workers
-        var workers = (c.nodes === 'cpus') ?
-            os.cpus().length : c.nodes;
+        var workers = (c.nodes === 'cpus') ? os.cpus().length : c.nodes;
         var new_workers = [];
         for (var i=0; i<workers; i++) {
             new_workers.push(cluster.fork({ CLUSTER_MASTER_PID: process.pid }));
@@ -362,7 +347,7 @@ Server.init_master_respond = function (retval, msg) {
                     CLUSTER_MASTER_PID: process.pid
                 });
                 new_worker.send({
-                    event: 'outbound.load_pid_queue', data: worker.process.pid
+                    event: 'outbound.load_pid_queue', data: worker.process.pid,
                 });
             }
         });
@@ -373,6 +358,8 @@ Server.init_child_respond = function (retval, msg) {
     switch (retval) {
         case constants.ok:
         case constants.cont:
+            Server.setup_http_listeners();
+            plugins.run_hooks('init_http', Server);
             return;
     }
 
@@ -408,15 +395,41 @@ Server.listening = function () {
     Server.ready = 1;
 };
 
-Server.get_http_docroot = function () {
-    if (Server.http_cfg.docroot) return Server.http_cfg.docroot;
+Server.init_http_respond = function () {
+    logger.loginfo('init_http_respond');
 
-    Server.http_cfg.docroot = path.join(
+    var WebSocketServer;
+    try { WebSocketServer = require('ws').Server; }
+    catch (e) {
+        logger.logerror('unable to load ws.\ndid you: npm install -g ws?');
+        return;
+    }
+
+    if (!WebSocketServer) {
+        logger.logerror('ws failed to load');
+        return;
+    }
+
+    Server.http.wss = new WebSocketServer({ server: Server.http.server });
+    logger.loginfo('Server.http.wss loaded');
+
+    plugins.run_hooks('init_wss', Server);
+};
+
+Server.init_wss_respond = function () {
+    logger.loginfo('init_wss_respond');
+    // logger.logdebug(arguments);
+};
+
+Server.get_http_docroot = function () {
+    if (Server.http.cfg.docroot) return Server.http.cfg.docroot;
+
+    Server.http.cfg.docroot = path.join(
         (process.env.HARAKA || __dirname),
         '/html'
     );
-    logger.loginfo('using html docroot: ' + Server.http_cfg.docroot);
-    return Server.http_cfg.docroot;
+    logger.loginfo('using html docroot: ' + Server.http.cfg.docroot);
+    return Server.http.cfg.docroot;
 };
 
 Server.handle404 = function(req, res){

--- a/tests/server.js
+++ b/tests/server.js
@@ -5,6 +5,7 @@
 
 function _set_up(done) {
     this.server = require('../server');
+    this.config = require('../config');
 
     done();
 }
@@ -30,7 +31,9 @@ exports.get_listen_addrs = {
     },
     'IPv4, custom port' : function (test) {
         test.expect(1);
-        var listeners = this.server.get_listen_addrs({listen: '127.0.0.1'}, 250);
+        var listeners = this.server.get_listen_addrs({
+            listen: '127.0.0.1'
+        }, 250);
         test.deepEqual(['127.0.0.1:250'], listeners);
         test.done();
     },
@@ -54,19 +57,25 @@ exports.get_listen_addrs = {
     },
     'IPv4 & IPv6 fully qualified' : function (test) {
         test.expect(1);
-        var listeners = this.server.get_listen_addrs({listen: '127.0.0.1:25,[::1]:25'});
+        var listeners = this.server.get_listen_addrs({
+            listen: '127.0.0.1:25,[::1]:25'
+        });
         test.deepEqual(['127.0.0.1:25','[::1]:25'], listeners);
         test.done();
     },
     'IPv4 & IPv6, default port' : function (test) {
         test.expect(1);
-        var listeners = this.server.get_listen_addrs({listen: '127.0.0.1:25,[::1]'});
+        var listeners = this.server.get_listen_addrs({
+            listen: '127.0.0.1:25,[::1]'
+        });
         test.deepEqual(['127.0.0.1:25','[::1]:25'], listeners);
         test.done();
     },
     'IPv4 & IPv6, custom port' : function (test) {
         test.expect(1);
-        var listeners = this.server.get_listen_addrs({listen: '127.0.0.1,[::1]'}, 250);
+        var listeners = this.server.get_listen_addrs({
+            listen: '127.0.0.1,[::1]'
+        }, 250);
         test.deepEqual(['127.0.0.1:250','[::1]:250'], listeners);
         test.done();
     },
@@ -114,5 +123,17 @@ exports.get_smtp_server = {
             test.equal(0, server.connections);
             test.done();
         }
+    },
+};
+
+
+exports.get_http_docroot = {
+    setUp : _set_up,
+    tearDown : _tear_down,
+    'gets a fs path': function (test) {
+        test.expect(1);
+        var path = this.server.get_http_docroot();
+        test.ok(path);
+        test.done();
     },
 };

